### PR TITLE
try to handle some 404 urls if possible

### PR DESCRIPTION
**Pull request by wluo**
see https://github.com/GigaOM/gigaom/issues/4854

### CODE REVIEW

File: tests/class-go-404-test.php [hunk 1]
```diff
@@ -0,0 +1,22 @@
+<?php
+/**
+ * GO_Marketo unit tests
+ */
+
+require_once dirname( __DIR__ ) . '/go-404.php';
+
+/**
+ * Our tests depend on the real Marketo authentication info from the system
```
Comment (misterbisson) [inline tests/class-go-404-test.php]: old

File: components/class-go-404.php [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+ * @author Gigaom <support@gigaom.com>
+ */
+class GO_404
+{
+	/**
+	 * constructor
+	 */
+	public function __construct()
+	{
+		add_action( 'template_redirect', array( $this, 'template_redirect' ) );
```
Comment (misterbisson) [inline components/class-go-404.php]: We should run this early, maybe on priority 3.

Comment (misterbisson) [inline components/class-go-404.php]: Crap, no. [Canonical redirects happen at the default priority](https://core.trac.wordpress.org/browser/tags/3.9.1/src/wp-includes/canonical.php#L530), and we don't want to precede those. Maybe this should go to 11?

File: components/class-go-404.php [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	/**
+	 * attempt to clean up and validate $url.
+	 *
+	 * @param string $url an url to be cleaned up and validated
+	 * @return a cleaned and validated url, or NULL if $url cannot be cleaned
+	 *  or validated
+	 */
+	public function clean_and_validate_url( $url )
+	{
+		$url = preg_replace( '!/[^/]*?$!', '', $url );
```
Comment (methnen) [inline components/class-go-404.php:46]: OCD me wants a line of white space here and below.

File: components/class-go-404.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+		if ( empty( $url ) )
+		{
+			return NULL;
+		}
+
+		$post_id = url_to_postid( $url );
```
Comment (methnen) [inline components/class-go-404.php:52]: OCD

File: components/class-go-404.php [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+/**
+ * The singleton
+ */
+function go_404()
+{
+	global $go_404;
+
+	if ( ! isset( $go_404 ) )
+	{
+		$go_404 = new GO_404();
```
Comment (methnen) [inline components/class-go-404.php:71]: don't need the `()` do we?

**Comment by methnen** [comment]
Minor nitpicks otherwise:

:cactus: :dash:

File: components/class-go-404.php (hunk 5 above)
Comment (misterbisson) [inline components/class-go-404.php:71]: We don't, but we've been including it anyway.

Comment (methnen) [inline components/class-go-404.php:71]: Ah, I've been leaving them out most of the time when I wasn't passing anything to the class.  Guess I'll switch my behavior if that's our thing.

File: components/class-go-404.php (hunk 3 above)
Comment (misterbisson) [inline components/class-go-404.php:46]: See https://github.com/GigaOM/go-404/pull/2

**Pull request merged**